service/history/api/respondworkflowtaskcompleted/api.go

```diff
@@ -429,26 +429,26 @@ func (handler *WorkflowTaskCompletedHandler) Invoke(
 			wtFailedShouldCreateNewTask ||
 			bufferedEventShouldCreateNewTask ||
 			activityNotStartedCancelled {
+
 			newWorkflowTaskType = enumsspb.WORKFLOW_TASK_TYPE_NORMAL
 
+		} else if updateRegistry.HasOutgoingMessages(true) {
 			// There shouldn't be any sent updates in the registry because
 			// all sent but not processed updates were rejected by server.
 			// Therefore, it doesn't matter if to includeAlreadySent or not.
-		} else if updateRegistry.HasOutgoingMessages(true) {
-			if completedEvent == nil || ms.GetNextEventID() == completedEvent.GetEventId()+1 {
-				newWorkflowTaskType = enumsspb.WORKFLOW_TASK_TYPE_SPECULATIVE
-			} else {
-				newWorkflowTaskType = enumsspb.WORKFLOW_TASK_TYPE_NORMAL
```

Comment:
Are we sure that this `else` block is never ever hit?  since the task type is `WORKFLOW_TASK_TYPE_UNSPECIFIED` otherwise.

Reply:
It is `else` for inner `if` not `else if` block. Similar to the change on API side, I now create new WFT always as speculative. Yes, `newWorkflowTaskType` can be `WORKFLOW_TASK_TYPE_UNSPECIFIED` and it means that no new WFT needs to be created.